### ISSUE TITLE
🐛 Bug Report: `BackendPluginManager` watching too many files

### ISSUE THREAD

**Issue by davidfestal** [backstage/backstage#20644]
### 📜 Description

The `backend-plugin-manager` watches for file changes in the dynamic plugins root folder.
However when watching for changes, it doesn't limit the depth, so that with many dynamic plugins having private dependencies, watching can produce the following errors:

```
2023-10-13T14:48:21.609Z backstage error error while watching '/opt/app-root/src/dynamic-plugins-root' ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js' errno=-28 syscall=watch code=ENOSPC path=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js filename=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js stack=Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21)
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21) {
  errno: -28,
  syscall: 'watch',
  code: 'ENOSPC',
  path: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js',
  filename: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
}
```
We should limit the folder depth where the `backend-plugin-manager` is watching for file changes. 

### 👍 Expected behavior

It should watch the changes without producing any error.

### 👎 Actual Behavior with Screenshots

```
2023-10-13T14:48:21.609Z backstage error error while watching '/opt/app-root/src/dynamic-plugins-root' ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js' errno=-28 syscall=watch code=ENOSPC path=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js filename=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js stack=Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21)
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21) {
  errno: -28,
  syscall: 'watch',
  code: 'ENOSPC',
  path: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js',
  filename: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
}
```

### 👟 Reproduction steps

1. setup the `backend-plugin-manager` to support dynamic backend plugins (see readme)
2. Add quite a number of dynamic plugins in the dynamic plugins root with a great number of private dependencies
3. See the error happening from times to times in the logs, according to the value of the `fs.inotify.max_user_watches` on a the system.

### 🖥️ Your Environment

Error happens on Linux.

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by tudi2d** [comment]
Thanks for reporting this @davidfestal! Assigning this to you - feel free to go ahead with a PR :)

**Comment by davidfestal** [comment]
> Thanks for reporting this @davidfestal! Assigning this to you - feel free to go ahead with a PR :)

That's planned already. Just have to find the time :-)

**Comment by davidfestal** [comment]
> > Thanks for reporting this @davidfestal! Assigning this to you - feel free to go ahead with a PR :)
> 
> That's planned already. Just have to find the time :-)

Done in PR #20755

**Comment by davidfestal** [comment]
Fixed in PR #20755
